src/components/AR/AR-session.ts

```diff
@@ -0,0 +1,97 @@
+import { ARProvider } from "./AR-provider";
+
+/**
+ * ARSession - master control for the AR session.
+ * - loads dependencies (aka providers)
+ * - handles global callbacks when AR session is started, ended
+ * - can end any running AR session.
+ * - renders AR button when the scene is loaded 
+ * 
+ * TODO - refactor checkSceneLoadProgress whenever we can control when the WL.onSceneLoaded is fired.
+ * Currently we listen to WL.onSceneLoaded AND to each registered provide's loading to decide when to show the AR button.
+ * Instead, we want to show the engines native 'loading' screen until all the providers are loaded  
+ */
+abstract class ARSession {
+  // tracking provider is basically a lib which has some tracking capabilities, so device native webXR, 8thwall, mind-ar-js, etc
+  private static trackingProviders: Array<ARProvider> = [];
+
+  // current running provider
+  private static currentTrackingProvider: ARProvider | null = null;
+
+  public static readonly onARSessionReady: Array<() => void> = [];
+
+  public static readonly onSessionStarted: Array<(trackingProvider: ARProvider) => void> = [];
+  public static readonly onSessionEnded: Array<(trackingProvider: ARProvider) => void> = [];
+
+  private static sceneHasLoaded = false;
+  private static _arSessionIsReady = false;
+
+  public static get arSessionReady  () {
+    return this._arSessionIsReady;
+  }
+
+  static {
+    if (window.document) {
+      WL.onSceneLoaded.push(() => {
```

Comment:
This is not going to work anmymore. This code needs to move where the engine instance is, global `WL` is going away. Basically I would say this code should go into `start`:

```ts
start() {
    this.engine.onSceneLoaded
}
```

Reply:
Everything is moved to init() to work with this.engine.
We lost some preloading time for XR8 though..